gsy_e_sdk/setups/Asset_API_scripts/degrees_of_freedom.py

```diff
@@ -0,0 +1,138 @@
+        pass
+
+    def read_requirements(self):
+        """Load the JSON file containing the list of requirements for each asset."""
+        with open(
+            os.path.join(module_dir, "resources/requirements.json"),
+            "r",
+            encoding="utf-8",
+        ) as file:
+            self.degrees_of_freedom = json.load(file)
+
+    def post_bid(self):
+        """Post a bid to the exchange."""
+        for area_uuid, area_dict in self.latest_grid_tree_flat.items():
+            asset_info = area_dict.get("asset_info")
+            if not asset_info:
+                continue
+            required_energy = asset_info.get("energy_requirement_kWh")
+            if required_energy:
+                rate = 10
+                if area_dict["area_name"] in self.degrees_of_freedom:
+                    requirements = self.build_requirements_dict(
+                        rate, required_energy, area_dict["area_name"]
+                    )
+                    self.add_to_batch_commands.bid_energy_rate(
+                        asset_uuid=area_uuid,
+                        rate=rate,
+                        energy=required_energy,
+                        requirements=requirements,
+                        replace_existing=True,
+                    )
+
+                else:
+                    self.add_to_batch_commands.bid_energy_rate(
+                        asset_uuid=area_uuid, rate=rate, energy=required_energy
+                    )
+
+    def build_requirements_dict(self, rate, energy, area_name):
+        """Return a dictionary with the requirements of the asset."""
+        asset_dof_list = self.degrees_of_freedom[area_name]
+        for asset_dof in asset_dof_list:
+            id_trading_partners_list = get_partner_ids(asset_dof["Trading Partners"])
+            energy_types_list = asset_dof["Energy Types"]
+            energy_requirement = (
+                energy
+                if asset_dof["Energy"] == "None"
+                else min(asset_dof["Energy"], energy)
+            )
+            rate_requirement = (
+                rate if asset_dof["Rate"] == "None" else asset_dof["Rate"]
+            )
+            requirements = [
+                {
+                    "trading_partners": id_trading_partners_list,
+                    "energy_type": energy_types_list,
+                    "energy": energy_requirement,
+                    "price": rate_requirement * energy_requirement,
+                }
+            ]
+        return requirements
+
+    def on_event_or_response(self, message):
+        pass
+
+    def on_finish(self, finish_info):
+        self.is_finished = True
+
+
+aggregator = Oracle(aggregator_name=ORACLE_NAME)
+asset_args = {"autoregister": True, "pubsub_thread": aggregator.pubsub}
+
+
+def get_partner_ids(partners):
+    """Get the trading partners' UUID from the assets' names."""
+    asset_ids = {v: k for k, v in asset_uuid_mapping.items()}
+    return [asset_ids[p] for p in partners]
```

Comment:
This only works because `asset_uuid_mapping` is defined before we run this function (`get_partner_ids()`). However, `asset_uuid_mapping` is not defined inside of this function, so if we refactor and move stuff around, or if we change the variable name, this method will break.
In theory, `asset_uuid_mapping` should be added to the function parameters:
```suggestion
def get_partner_ids(partners, asset_uuid_mapping):
    """Get the trading partners' UUID from the assets' names."""
    asset_ids = {v: k for k, v in asset_uuid_mapping.items()}
    return [asset_ids[p] for p in partners]
```
However, this means that the `asset_uuid_mapping` should also be available inside of the `Oracle` instance (because it needs to be used by `build_requirements_dict()`). But the Oracle instance is used inside of `register_asset_list()` (which generates `asset_uuid_mapping`) so we're in a loop. This would require a bit more refactoring IMO, as it's probably not an immediate fix.